### ISSUE TITLE
Write guidance on how to launch (non apphost) apps

### ISSUE THREAD

**Issue by richlander** [dotnet/runtime#88754]
There are scenarios where apps need to be launched with `dotnet` host. The big question is how to find the host, with the two obvious places to look being:

- `PATH`
- apphost configuration (`DOTNET_ROOT`, `/etc/dotnet/install_location`, ...)

I don't believe we've documented this.

This question may be similar to how one locates `node` from code given the use of `nvm`. I haven't looked at that pattern/guidance, but it seems similar.

**Comment by jaredpar** [comment]
This comes up frequently in the C# compiler source code as there are a number of cases where we need to launch a managed process. Effectively we have `csc.dll` and need to launch it as a process but that requires us first finding a `dotnet` to use. The pattern for finding `dotnet` hasn't been documented and as a result our implementation has been derived by looking at ENV variables set by the runtime, reviewing patterns in other parts of the .NET SDK, etc ... Basically trying to draw from other examples. 

There are generally two scenarios I think about in this area:

1. How the C# compiler is executed in the .NET SDK. Essentially what should be the experience when the customer types `dotnet build`? 
2. How the C# compiler is executed within our own infrastructure. There is a strong incentive for our own infra to be able to control the `dotnet` that is used when running tools in the SDK. For example the `runtime` repo uses this to effectively dogfood the latest runtime against SDK tools. This is very valuable.

After thinking this through this morning I think the right approach going forward is to **only** consult `$PATH`. Basically find the first `dotnet` instance and use that. This mostly closely mirrors how a customer would run our own tooling and is most likely to line up with their expectations. Basically `dotnet build` uses the same host as `dotnet exec csc`. 

That would have the following impact on our product and testing:

1. Our product code which today consults `$DOTNET_ROOT` or `$DOTNET_HOST_PATH` for the purpose of finding `dotnet` should cease doing so. Instead consult `$PATH` only. 
2. Our infra code which was trying to insert a `dotnet` for testing / dogfood purposes by changing values like `$DOTNET_ROOT` should cease doing so. Instead that specific infra run should prepend the directory that contains the recently built `dotnet` to `$PATH`

**Comment by jaredpar** [comment]
If you're asking what is `$DOTNET_HOST_PATH` I had the same question. In researching through our code I _believe_ that it is effectively an arcade invention. It's a mechanism for us to control where arcade puts `dotnet` into our enlistments. Unfortunately this seems to have filtered into our actual production code as well. That feels wrong. 

Perhaps I have the history / intent of this variable wrong and if so please correct me 😄

**Comment by vitek-karas** [comment]
I think `DOTNET_HOST_PATH` is something SDK uses - there are many use cases in that repo: https://github.com/search?q=repo%3Adotnet%2Fsdk%20DOTNET_HOST_PATH&type=code
I think it uses it to pass the location of the host from the parent to the child processes, so that child processes use the same SDK/runtime install which was used to run the top level command. It's possible Arcade uses it as well. But I could be wrong.
For the usage inside SDK though we might want to keep looking at `DOTNET_HOST_PATH`. The problematic scenario is for example:
```
/my/private/install/dotnet build
```

This should spawn the child nodes for msbuild, csc and so on using the runtime/SDK from `/my/private/install` and not look for it on the `PATH`. For example if I have only installed 7.0 globally (so it's on PATH) but then I run `/my/private/8.0-preview.7/dotnet` then things should just work.

Unless we also change SDK to modify PATH for its child processes.

Outside of SDK I agree that `PATH` is probably the best approach. Installers, scripts and users will set `PATH` to point to `dotnet.exe` because we tell them to and because they need it anyway (so that they can run `dotnet build` and similar). It will also be set in CI systems to point to the dotnet from the SDK they want used and so on.
`DOTNET_ROOT` in itself is not enough and there's much more involved algorithm how to find the runtime if that variable is not set. That's basically what `nethost` implements.

**Comment by jaredpar** [comment]
> I think DOTNET_HOST_PATH is something SDK uses - 

I 100% agree it uses it, I'm questioning why it does that. The `$DOTNET_HOST_PATH` does not appear to be a variable that we support as a product. The only places I can find where it is set is through arcade, not the customer facing product. As such I'm questioning whether or not those uses in the SDK are correct and at the moment I don't believe they are. 

> For the usage inside SDK though we might want to keep looking at DOTNET_HOST_PATH.

Why though? I'm struggling to see what value it provides vs. always using `$PATH` and just modifying that to use it. 

If we do want a variable that special cases the SDK though then 

1. We need to universally use it. Today it's hit and miss from what I can see. 
2. Having it called `DOTNET_` feels wrong. This is for effectively our internal testing purposes. Giving it a name like `DOTNTE_` feels like we're elevating it to an actual .NET feature.

**Comment by vitek-karas** [comment]
> > For the usage inside SDK though we might want to keep looking at DOTNET_HOST_PATH.

> Why though? I'm struggling to see what value it provides vs. always using $PATH and just modifying that to use it.

I realized this as well after I sent the response. It would be simpler if SDK just modified `PATH` for its child processes I assume and then everything could use `PATH.

Unfortunately `PATH` is not great to use - especially on Windows. Historically our guidance (given to us by other teams) was to try to not parse `PATH` and make sense out of it. But maybe it's OK for just spawning processes, since we can let the OS do it for us I assume.

**Comment by jaredpar** [comment]
> Historically our guidance (given to us by other teams) was to try to not parse PATH and make sense out of it

Yeah I'm not really thrilled about parsing `$PATH` either. Ended up just stealing the code from [msbuild that does it](https://github.com/dotnet/msbuild/blob/main/src/Utilities/ToolTask.cs#L1259). This is essentially how our entire build tool stack works today so it seems like a good source to draw some. 

Would be nice to have an API for `FindOnPath` but that seems unlikely. Also I'd still have to polyfill that for downlevel targets.

**Comment by rainersigwald** [comment]
It's come up before but I couldn't find an issue so https://github.com/dotnet/msbuild/issues/9018.

**Comment by agocke** [comment]
@vitek-karas is there a managed API for finding the location of the muxer? Is this something we want to provide?

**Comment by vitek-karas** [comment]
No - we don't have an API like that. There's something similar in SDK: https://github.com/dotnet/sdk/blob/a4b77d4fd5e878fd42aef22d86c0668c90888da6/src/Resolvers/Microsoft.DotNet.MSBuildSdkResolver/MSBuildSdkResolver.cs. And there seem to be other versions of this elsewhere, for example: https://github.com/microsoft/MSBuildLocator/blob/master/src/MSBuildLocator/DotNetSdkLocationHelper.cs (this one even mentions that maybe it should use nethost but doesn't).
`dotnet test` also has something similar, where it is actually looking for `dotnet.exe` in some cases.

I know we've had several discussions with various SDK teams about this in the past, but no specific result. The problem is that there are like 5 versions of this code each with its own set of quirks and it's too late to change that.

**Comment by jaredpar** [comment]
There is one other view to consider which was brought up in https://github.com/dotnet/roslyn/issues/69023. Essentially all the tools that execute as part of an SDK action should use the same `dotnet` host. 

For example: 

```sh
> export PATH=/some/dotnet/install:$PATH
> which dotnet
/some/dotnet/install/dotnet
> /some/dotnet2/install/dotnet build 
```

In this view point tools like the compiler would be launched using `/some/dotnet2/install/dotnet` even though it is not the first `dotnet` on `$PATH`. Or possibly it's not on `$PATH` at all. 

I'm certainly sympathetic to this point of view. Mostly because the opposite view seems quite strange. Basically if I launch `msbuild` with one `dotnet` host and then suddenly other tools are launching with another `dotnet` host that is effectively revealing implementation details of the build. 

If there was an easy way from a process to find the `dotnet` host that created it then I'd lean towards this as the _primary_ mechanism for finding the `dotnet` host. That cannot be the sole mechanism for finding `dotnet` though as we still need to fall back to `$PATH` lookup in cases where our .NET core tools are loaded from .NET Framework processes.

**Comment by jaosnz-rep** [comment]
The original error mentioned that the picture is truncated is not reproduced at present, but when the window is resized on Windows and MacCatalyst, there will be a problem that the Image size doesn't match page size.

**Comment by agocke** [comment]
I think an API called something like `AppContext.ExecutingMuxerPath` makes sense -- the idea is that we would internally use the native hosting APIs to grab the location of hostfxr, and then from there we would try to find `dotnet.exe` (the muxer) in the expected location.

The problem being that there are a number of uncommon configurations where this will not work, including self-contained deployments where there just isn't a muxer to find. So this API would probably return `null` in a variety of cases where it's not possible to return a useful path.

@jaredpar Would this address all scenarios you're thinking of? I think the basic idea is that this would fix the problem, "I would like to start a new dotnet process using the same runtime/host that I'm running under." It wouldn't try to solve the problem of launching the "first" .NET process (either from native code or from the desktop framework). In those cases you would have to rely on PATH being appropriately configured, or use the native hosting APIs.

Another limitation would be that the muxer serves double-duty as the entry point to both the runtime and to the SDK. There's no guarantee that an SDK will actually be present in any given dotnet process though. So people have to be careful to only use this to execute DLLs, not run SDK commands.

**Comment by xoofx** [comment]
> For the usage inside SDK though we might want to keep looking at DOTNET_HOST_PATH.
> > Why though? I'm struggling to see what value it provides vs. always using $PATH and just modifying that to use it.

I'm also not a fan of overriding PATH and looking into it, because it means that we can't use a dotnet exe without wrapping it with a batch/pwsh to setup this variable. That's really not great.

Even if `DOTNET_HOST_PATH` is not official, it seems that it is still considered and used all over the places as the primary location to look before going to the path, so why not making it official? Could we then always enforce it from dotnet.exe? (when run for the first time without a `DOTNET_HOST_PATH`?)

> I think an API called something like AppContext.ExecutingMuxerPath makes sense -- the idea is that we would internally use the native hosting APIs to grab the location of hostfxr, and then from there we would try to find dotnet.exe (the muxer) in the expected location.

Having an API would be nice, but I would think that making `DOTNET_HOST_PATH` first class would be simpler and would even make it working with launching intermediate sub-child process shells that inherit from this variable and can resolve to dotnet.exe without having to explicitly propagate this variable around. It makes it more consistent to stick with an environment variable to propagate such things between managed and native processes.

**Comment by jaredpar** [comment]
> "I would like to start a new dotnet process using the same runtime/host that I'm running under.

For me at least that is the primary problem to solve. Basically I have a suite of tools that launch each other and I want them to use the same host.

**Comment by jaredpar** [comment]
> Having an API would be nice, but I would think that making DOTNET_HOST_PATH first class would be simpler 

This is a proven approach we've used in MsBuild and it functions well for this type of scenario. Perhaps this new API just seeds this value more automatically.

**Comment by agocke** [comment]
@elinor-fung Any thoughts on the above? Either a managed API, or setting `DOTNET_HOST_PATH` environment variable when initializing the runtime?

**Comment by elinor-fung** [comment]
> setting DOTNET_HOST_PATH environment variable when initializing the runtime

I don't really want to get the host itself into the business of setting environment variables. I think it would be reasonable to enable something like `AppContext.GetData("DOTNET_HOST_PATH")`.

> > "I would like to start a new dotnet process using the same runtime/host that I'm running under.
> 
> For me at least that is the primary problem to solve. Basically I have a suite of tools that launch each other and I want them to use the same host.

Is it the same host that is desired, or the same runtime? Given https://github.com/dotnet/designs/pull/303, those could be different locations. My impression is that the same runtime would be the desire.

**Comment by rainersigwald** [comment]
For build scenarios I _think_ we want "same runtime and SDK as current MSBuild process"--at least for SDK built-in tooling like Roslyn. Tools packaged in NuGet packages _may_ want more flexibility but that feels like a good default.

**Comment by xoofx** [comment]
> I don't really want to get the host itself into the business of setting environment variables

On that topic, I have been having a difficult time to track down which process/main is actually setting `DOTNET_HOST_PATH`? When we are referring to `DOTNET_HOST_PATH`, is it acknowledged that it is only about the muxer? (`dotnet` exe, in the SDK situation)

**Comment by xoofx** [comment]
> For build scenarios I _think_ we want "same runtime and SDK as current MSBuild process"--at least for SDK built-in tooling like Roslyn. Tools packaged in NuGet packages _may_ want more flexibility but that feels like a good default.

By curiosity, how do we propagate from process to process the same runtime? (assuming we go through the `dotnet` muxer?) Do we point to the `dotnet path/sdk/x.y.z/dotnet.dll` in a specific SDK version folder?

**Comment by jaredpar** [comment]
> I don't really want to get the host itself into the business of setting environment variables. I think it would be reasonable to enable something like AppContext.GetData("DOTNET_HOST_PATH").

I understand, and agree with, the reluctance for the runtime to be setting environment variables. At the same time though it would seem odd if the runtime is giving an API to find the host executable but then our SDK tool chain is effectively inventing another mechanism via `$DOTNET_HOST_PATH` environment variable. If we go with a new API then I think we need to consider moving the SDK to using that mechanism as the primary mode.

> Is it the same host that is desired, or the same runtime?

Agree it's the same runtime. Basically the location we can `dotnet exec` with to get the same runtime executing the current process.

**Comment by agocke** [comment]
> Is it the same host that is desired, or the same runtime? Given https://github.com/dotnet/designs/pull/303, those could be different locations. My impression is that the same runtime would be the desire.

Actually, I don't think my statement was well-defined. What DOTNET_HOST_PATH is intended to do is point to a muxer that can run the target binary. There's no guarantee that binary will end up actually using the same runtime, due to all the decisions that go happen when choosing the binary.

I think my proposal is: find the muxer associated with the current host, if one exists. Run under that muxer. No guarantees about SDK or runtime.

**Comment by agocke** [comment]
> I understand, and agree with, the reluctance for the runtime to be setting environment variables. At the same time though it would seem odd if the runtime is giving an API to find the host executable but then our SDK tool chain is effectively inventing another mechanism via `$DOTNET_HOST_PATH` environment variable. If we go with a new API then I think we need to consider moving the SDK to using that mechanism as the primary mode.

I think there are potentially two different scenarios here: do you want tools launched under your process to know that this is a .NET process and where it was launched from? Or do you simply want managed processes to be able to find the current runtime?

I think of this as the divide between native children and managed children. If you want native children to know the location of the host, you probably want to use an environment variable. But also, I don't think this is an appropriate contract to use for the .NET runtime -- we wouldn't want to automatically pollute the environment variable space of child processes.

Conversely, a managed API is effectively an entirely internal contract to a .NET binary. That's definitely the kind of thing that makes sense to build into every .NET app.

So I think an AppContext switch makes a lot of sense as a universal host feature, and environment variable less so. But, I would also find it reasonable if the `dotnet` CLI/SDK, in particular, wanted to have all child processes inherit a .NET-specific variable. That would be something unique to running tools under the SDK and we wouldn't expect, say, a random WinForms app to start passing its host path to native child processes.

**Comment by zivkan** [comment]
FWIW, I just lost over a day of productivity, which could have been avoided if there was an API to allow apps to launch child processes with the same dotnet muxer as the parent process.

I'm creating a new command for the dotnet SDK, and the dotnet SDK sets the DOTNET_HOST_PATH environment variable which our product code uses. But I'm working in the NuGet.Client repo, so integration tests can't run the command as a child process until after the new command is inserted into the dotnet sdk repo. So, I used a console app as a temporary alternative to running the final dotnet command. But in order to solve `Environment.GetEnvironmentVariable`'s nullable warning, my first mistake was to fallback to `dotnet` on the PATH. Though a comedy of several other errors that is not relevant to this issue, it took me hours to figure out and solve, finally setting the environment variable in the test.

**Comment by elinor-fung** [comment]
Here's what I'm thinking for using `AppContext` data - `AppContext.GetData("DOTNET_HOST_PATH")` would be a string value:
- Absolute path pointing to the `dotnet` executable corresponding to the runtime being used to run the application
- Not necessarily the same as the running process or the `dotnet` for the loaded `hostfxr`:
  - with local SDK paths, the running `dotnet` and `hostfxr` may not correspond to the runtime used to run the SDK .dll
  - when using `apphost`, the `apphost` is the executing process, but the property points at the `dotnet` corresponding to the runtime that was resolved for the app. 
- Not set for scenarios where there is no corresponding `dotnet` for the runtime - for example: self-contained, NativeAOT, using a .NET runtime install that (for whatever reason) doesn't have a `dotnet` executable
- No guarantee that an app launched through that `dotnet` will actually use the same runtime and frameworks versions
  - runtime version for the current app will be available to that `dotnet`, along with any other versions under that install
  - whether or not that same version is used depends on configuration (target version, roll forward) of what is being launched
- No guarantee that any SDK is available to that `dotnet` - unless the original app is the SDK's `dotnet.dll` itself

Would that make sense / actually be useful for folks here?

**Comment by jkotas** [comment]
With this proposal, is it going to be possible to replace https://github.com/dotnet/sdk/blob/ea7c246d30be4db27c9d265afef7162dce8d3ed1/src/Cli/Microsoft.DotNet.Cli.Utils/Muxer.cs#L39-L77 in .NET SDK with just `_muxerPath = AppContext.GetData("DOTNET_HOST_PATH")`?

**Comment by elinor-fung** [comment]
As it exists now, I believe so. With https://github.com/dotnet/sdk/issues/49667 for shipping the dotnet CLI as a native AOT application, if the native AOT SDK needs to launch a regular .NET app, I think it would still need some fallback of looking relative to itself.

@jaredpar @rainersigwald @baronfel - please correct me if I'm missing something.

**Comment by baronfel** [comment]
I wrote a doc today based on some findings that echo what folks here are saying I think: https://hackmd.io/pKvDHW89TwOKPbG7FvnLWw

I think we need more shared components similar to the MSBuild SdkResolvers, but for general IDE/tooling scenarios outside of MSBuild. @elinor-fung I'd especially like your eyes on that^

**Comment by baronfel** [comment]
> As it exists now, I believe so. With [dotnet/sdk#49667](https://github.com/dotnet/sdk/issues/49667) for shipping the dotnet CLI as a native AOT application, if the native AOT SDK needs to launch a regular .NET app, I think it would still need some fallback of looking relative to itself.
> 
> [@jaredpar](https://github.com/jaredpar) [@rainersigwald](https://github.com/rainersigwald) [@baronfel](https://github.com/baronfel) - please correct me if I'm missing something.

Yeah, in that case we'd still need to trampoline but @agocke has already shown a way forward there. I'm very happy with this AppContext approach 👍
